### PR TITLE
Issue/2130 headline

### PR DESCRIPTION
**Pull request by tatjana-gartner2**
### Pull Request Checklist

- [x] I have read and accept the [Contributor License Agreement](https://github.com/porscheofficial/oss-docs/blob/67c5b24838a293ce7a964884e6005eb71f2b8579/CONTRIBUTOR_LICENSE_AGREEMENT.md)

#### References

- Preview: https://designsystem.porsche.com/issue/.../

#### Scope

Short description of the pull request scope

### CODE REVIEW

File: packages/components-angular/src/app/examples/table-example-advanced.component.ts [hunk 1]
```diff
@@ -6,7 +6,7 @@ import { dataAdvanced, DataAdvanced, headAdvanced } from '@porsche-design-system
   selector: 'page-table-example-advanced',
   template: `
     <p-table (sortingChange)="onSortingChange($event)">
-      <p-headline slot="caption" variant="headline-3">Some visual caption</p-headline>
+      <p-heading slot="caption" variant="headline-3">Some visual caption</p-heading>
```
Comment (denyo) [inline packages/components-angular/src/app/examples/table-example-advanced.component.ts:9]: `variant="headline-3"` doesn't seem right?

File: packages/components-js/tests/e2e/puppeteer/specs/heading.e2e.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+  });
+});
+
+describe('lifecycle', () => {
+  it('should work without unnecessary round trips on init', async () => {
+    await initHeading({ variant: 'heading-1' });
+    const status = await getLifecycleStatus(page);
+
+    expect(status.componentDidLoad['p-heading'], 'componentDidLoad: p-heading').toBe(1);
+    expect(status.componentDidUpdate['p-heading'], 'componentDidUpdate: p-heading').toBe(0);
```
Comment (denyo) [inline packages/components-js/tests/e2e/puppeteer/specs/heading.e2e.ts:91]: This is redundant with the check for `componentDidUpdate.all`.

File: packages/components-js/tests/e2e/puppeteer/specs/heading.e2e.ts [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    await initHeading({ variant: { base: 'small', l: 'large' } });
+    const status = await getLifecycleStatus(page);
+
+    expect(status.componentDidLoad['p-heading'], 'componentDidLoad: p-heading').toBe(1);
+
+    expect(status.componentDidLoad.all, 'componentDidLoad: all').toBe(1);
+    expect(status.componentDidUpdate.all, 'componentDidUpdate: all').toBe(0);
+  });
+
+  it('should work without unnecessary round trips after state change', async () => {
```
Comment (denyo) [inline packages/components-js/tests/e2e/puppeteer/specs/heading.e2e.ts:107]: It's a prop change or where is there any state?

File: packages/components-js/tests/e2e/puppeteer/specs/heading.e2e.ts [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    `
+    <p-heading ${attrs}>
+      ${slot ? slot : 'Some Heading'}
+    </p-heading>`
+  );
+};
+
+const getHost = () => selectNode(page, 'p-heading');
+
+const getHeadingTagName = (): Promise<string> =>
```
Comment (denyo) [inline packages/components-js/tests/e2e/puppeteer/specs/heading.e2e.ts:35]: Could be refactored to something like `(await getHost()).evaluate(el => el.shadowRoot.querySelector('.root').tagName)`

File: packages/components/scripts/generateDSRComponents.ts [hunk 5]
```diff
@@ -193,10 +193,10 @@ import { get${componentName}Css } from '${stylesBundleImportPath}';
         .replace(/this\.props\.getAttribute\('tabindex'\)/g, 'null') // button
         .replace(/getTextListItemCss\(listType, orderType, isNestedList\)/, "''") // text-list-item
         .replace(
-          /(getHeadlineTagName|getHTMLElement|getClosestHTMLElement|getDirectChildHTMLElement)\(this\.props/,
+          /(getHeadingTagName|getHTMLElement|getClosestHTMLElement|getDirectChildHTMLElement)\(this\.props/,
           '$1(null'
         ) // headline, text, text-list, tag
-        .replace(/ = getHeadlineTagName/, ': any$&') // headline
+        .replace(/ = getHeadingTagName/, ': any$&') // headline
```
Comment (denyo) [inline packages/components/scripts/generateDSRComponents.ts:199]: Comments don't make no more sense.

File: packages/components/src/components/heading/heading-styles.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  variant: HeadingVariant,
+  align: TextAlign,
+  color: Extract<TextColor, 'primary' | 'default' | 'inherit'>,
+  ellipsis: boolean,
+  theme: Theme
+): string => {
+  return getCss({
+    '@global': {
+      ':host': {
+        display: 'block',
```
Comment (denyo) [inline packages/components/src/components/heading/heading-styles.ts:68]: hidden styles are missing, important probably, too.

File: packages/components/src/components/heading/heading-styles.ts [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+  headingXXLargeStyle,
+  fontWeight,
+  textSmallStyle,
+  displayLargeStyle,
+} from '@porsche-design-system/utilities-v2';
+import { getEllipsisJssStyle, getSlottedTypographyJssStyle } from '../../styles/typography-styles';
+import { isVariantType } from './heading-utils';
+
+const headingMap: Record<VariantType, JssStyle> = {
+  'large-title': displayLargeStyle,
```
Comment (denyo) [inline packages/components/src/components/heading/heading-styles.ts:19]: Can't we use `HEADING_VARIANTS` here somehow instead of redefining everything?

File: packages/components/src/components/heading/heading.tsx [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+  TEXT_ALIGNS,
+  validateProps,
+} from '../../utils';
+import type { PropTypes, TextAlign, TextColor, Theme } from '../../types';
+import type { HeadingTag, HeadingVariant } from './heading-utils';
+import { getHeadingTagName, HEADING_TAGS } from './heading-utils';
+import { getComponentCss } from './heading-styles';
+
+const propTypes: Omit<PropTypes<typeof Heading>, 'variant'> = {
+  // variant: AllowedTypes.string, // TODO: with all the different values this can't easily be validated
```
Comment (denyo) [inline packages/components/src/components/heading/heading.tsx:16]: Shouldn't this be cleaned up / reduced with the new component? Currently it seems just like a rename.

File: packages/components/src/components/headline/headline-styles.ts [hunk 9]
```diff
@@ -1,85 +1,12 @@
...
-};
-
-export const getSlottedCss = (host: HTMLElement): string => {
-  return getCss(
-    buildSlottedStyles(
-      host,
-      mergeDeep(getBaseSlottedStyles({ withDarkTheme: true }), { '& a': { textDecoration: 'none' } })
-    )
-  );
+  return getHeadingHeadlineStyles(variant, align, color, ellipsis, theme);
```
Comment (denyo) [inline packages/components/src/components/headline/headline-styles.ts:11]: Could probably just reexport the function instead of wrapping it in another function?

File: packages/components/src/components/heading/heading-utils.ts [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+import { getHTMLElement } from '../../utils';
+
+export const HEADING_VARIANTS = [
+  'large-title',
+  'heading-1',
+  'heading-2',
+  'heading-3',
+  'heading-4',
+  'heading-5',
+  'headline-1', // deprecated
```
Comment (denyo) [inline packages/components/src/components/heading/heading-utils.ts:11]: Hmm, now the old and the new component got bigger and even more flexible? Is this really intended?

File: packages/components/src/components/heading/heading.code.md [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+
+@Component
+export default class Code extends Vue {
+  config = { themeable: true };
+
+  customVariant = "{ base: 'small', l: 'medium' }";
+  color = 'default';
+  align = 'center';
+
+  variant = HEADING_VARIANTS.map((item) => `<p-heading variant="${item}">${sentence}</p-heading>`).join('\n');
```
Comment (denyo) [inline packages/components/src/components/heading/heading.code.md:113]: This produces live examples for the deprecated stuff.

File: packages/components/scripts/cleanReadmes.ts [hunk 12]
```diff
@@ -22,7 +22,7 @@ const adjustHeadline = (str: string): string => {
   const h1Replacements: { [key in TagName]?: string } = {
     'p-flex': 'Flex',
     'p-grid': 'Grid',
-    'p-headline': 'Typography',
+    'p-heading': 'Typography',
```
Comment (denyo) [inline packages/components/scripts/cleanReadmes.ts:25]: I think the table of contents is missing `p-headline` here: https://designsystem.porsche.com/issue/2130/components/typography/props

In addition the possible values for `variant` don't seem to be correct.

File: packages/components/src/components/headline/headline.code.md [hunk 13]
```diff
@@ -110,7 +117,7 @@ export default class Code extends Vue {
   color = 'default';
   align = 'center';
 
-  variant = HEADLINE_VARIANTS.map((item) => `<p-headline variant="${item}">${sentence}</p-headline>`).join('\n');
+  variant = HEADING_VARIANTS.map((item) => `<p-headline variant="${item}">${sentence}</p-headline>`).join('\n');
```
Comment (denyo) [inline packages/components/src/components/headline/headline.code.md:120]: Even more examples with new possible but deprecated values here, too.

File: packages/components/src/utils/tag-name.ts [hunk 14]
```diff
@@ -13,10 +13,12 @@ export const getTagNameWithoutPrefix = (host: HTMLElement): TagName => {
   return (tagNameWithoutPrefix || tagName) as TagName; // return tagName as fallback for default tags
 };
 
-// prevent internal usage of p-headline and p-text
-type AllowedTagNameCamelCase = Exclude<TagNameCamelCase, 'pHeadline' | 'pText'>;
+// prevent internal usage of p-heading and p-text
```
Comment (denyo) [inline packages/components/src/utils/tag-name.ts:16]: headline is missing now?

File: packages/components/src/utils/validation/throwIfComponentIsDeprecated.ts [hunk 15]
```diff
@@ -0,0 +1,5 @@
+import { getTagName } from '../../utils';
+
+export const throwIfComponentIsDeprecated = (host: HTMLElement, message: string): void => {
```
Comment (denyo) [inline packages/components/src/utils/validation/throwIfComponentIsDeprecated.ts:3]: That's not throwing anything, also what's the benefit of wrapping `console.warn` in another function that basically combines 2 parameters with a `:`?

File: packages/components/src/utils/validation/throwIfComponentIsDeprecated.spec.ts [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+
+describe('with host element', () => {
+  it('should throw warning', () => {
+    const host = document.createElement('div');
+    jest.spyOn(console, 'warn').mockImplementation();
+    throwIfComponentIsDeprecated(
+      host,
+      'This component is deprecated and will be removed with next major release. Use "link" component with corresponding social icon instead.'
+    );
+    expect(console.warn).toHaveBeenCalledWith(expect.stringContaining(warningMessage));
```
Comment (denyo) [inline packages/components/src/utils/validation/throwIfComponentIsDeprecated.spec.ts:14]: Kind of unprecise because there could be other things before and after the string?

File: packages/storefront/src/pages/accessibility/statement.md [hunk 17]
```diff
@@ -98,14 +98,20 @@ We continually assess and work to ensure that our Web presence is in conformance
     margin-top: $pds-spacing-static-medium;
   }
 
+  p-text + p-heading[variant="heading-3"],
```
Comment (denyo) [inline packages/storefront/src/pages/accessibility/statement.md:101]: Do we still need all these selectors with `p-headline` now?

File: packages/storefront/src/pages/start-coding/react/advanced.md [hunk 18]
```diff
@@ -29,7 +29,7 @@ ReactDOM.render(
 );
 ```
 
-In the following example the `PHeadline` component will render as `<sample-prefix-p-headline>`.
+In the following example the `PHeadline` component will render as `<sample-prefix-p-heading>`.
```
Comment (denyo) [inline packages/storefront/src/pages/start-coding/react/advanced.md:32]: Mismatch in the sentence.

File: packages/storefront/src/pages/start-coding/angular/getting-started.md [hunk 19]
```diff
@@ -55,7 +55,7 @@ import { Component } from '@angular/core';
   selector: 'app-root',
   template: `
     <div id="app">
-      <p-headline variant="headline-1">Headline</p-headline>
+      <p-heading variant="heading-1">Headline</p-heading>
```
Comment (denyo) [inline packages/storefront/src/pages/start-coding/angular/getting-started.md:58]: Maybe rephrase.

File: packages/storefront/tests/e2e/fixtures/sitemap.json [hunk 20]
```diff
@@ -162,6 +162,8 @@
   "/components/textarea/props",
   "/components/textarea/usage",
   "/components/typography",
+  "/components/typography/heading",
+  "/components/typography/heading/",
```
Comment (denyo) [inline packages/storefront/tests/e2e/fixtures/sitemap.json:166]: Looks like a duplicate, perhaps align for consistent links.

File: packages/storefront/src/pages/utilities/scss/functions.md [hunk 21]
```diff
@@ -90,12 +90,12 @@ p {
 }
 ```
 
-### Headline
+### Heading
 
 Given values are:  
-`title-large | headline-1 | headline-2 | headline-3 | headline-4 | headline-5`
+`title-large | heading-1 | heading-2 | heading-3 | heading-4 | heading-5`
```
Comment (denyo) [inline packages/storefront/src/pages/utilities/scss/functions.md:96]: Do these utils even exist??

File: packages/components/src/utils/validation/throwIfComponentIsDeprecated.ts (hunk 15 above)
Comment (tatjana-gartner2) [reply packages/components/src/utils/validation/throwIfComponentIsDeprecated.ts]: used this function because it was used in link social already for deprecated warning

File: packages/components/rollup-styles.config.js [hunk 22]
```diff
@@ -18,12 +18,16 @@ const generateStylesEntryFile = () => {
   const stylesExports = stylesPaths
     .map((utilPath) => {
       const isCommonComponent = utilPath.includes('/components/common/');
-      const styleExport = isCommonComponent
-        ? `*`
-        : `{ getComponentCss as get${pascalCase(/([-a-z]+)-styles\.ts/.exec(utilPath)[1])}Css }`;
-      const fromPath = `./components${utilPath.replace(componentsDir, '').replace(/\.ts$/, '')}`;
+      const isHeadlineComponent = utilPath.includes('/components/headline/');
```
Comment (denyo) [inline packages/components/rollup-styles.config.js:21]: Seems to be the wrong place to have component specific tweaks in a build config.

File: packages/components/scripts/generateDSRComponents.ts [hunk 23]
```diff
@@ -80,7 +80,11 @@ const generateDSRComponents = (): void => {
           `import { Component } from 'react';
 import { minifyCss } from '../../minifyCss';
 import { stripFocusAndHoverStyles } from '../../stripFocusAndHoverStyles';
-import { get${componentName}Css } from '${stylesBundleImportPath}';
+${
+  componentName === 'Headline'
```
Comment (denyo) [inline packages/components/scripts/generateDSRComponents.ts:84]: Also seems to be wrong to have these changes now.

File: packages/components/src/components/heading/heading-styles.ts [hunk 24]
```diff
@@ -50,6 +58,7 @@ export const getComponentCss = (
   return getCss({
     '@global': {
       ':host': {
+        ...addImportantToEachRule({ ...hostHiddenStyles }),
```
Comment (denyo) [inline packages/components/src/components/heading/heading-styles.ts:61]: Useless object composition for hidden styles.

File: packages/components/src/components/headline/headline.tsx [hunk 25]
```diff
@@ -47,8 +49,11 @@ export class Headline {
   /** Adapts the text color depending on the theme. Has no effect when "inherit" is set as color prop. */
   @Prop() public theme?: Theme = 'light';
 
-  public connectedCallback(): void {
-    attachSlottedCss(this.host, getSlottedCss);
+  public componentWillLoad(): void {
+    warnIfComponentIsDeprecated(
```
Comment (denyo) [inline packages/components/src/components/headline/headline.tsx:53]: Please use `warnIfDeprecatedComponentIsUsed()` util.

File: packages/components/src/components/headline/headline.tsx [hunk 26]
```diff
@@ -2,20 +2,22 @@ import { Component, Element, h, Host, JSX, Prop } from '@stencil/core';
...
 } from '../../utils';
 import type { PropTypes, TextAlign, TextColor, Theme } from '../../types';
-import type { HeadlineTag, HeadlineVariant } from './headline-utils';
-import { getHeadlineTagName, HEADLINE_TAGS } from './headline-utils';
-import { getComponentCss, getSlottedCss } from './headline-styles';
+import type { HeadingTag } from '../heading/heading-utils';
+import type { HeadlineVariantDeprecated } from './headline-utils';
+import { getHeadlineTagName } from './headline-utils';
+import { HEADING_TAGS } from '../heading/heading-utils';
+import { getComponentCss } from '../heading/heading-styles';
```
Comment (denyo) [inline packages/components/src/components/headline/headline.tsx:16]: This is probably the root cause for all rollup and dsr component import related changes.
Could easily be solved similar as for link-social.

Comment (tatjana-gartner2) [reply packages/components/src/components/headline/headline.tsx:16]: Yes it is the root cause 🙈 so i guess its better to add a styles file again and wrap in another function

Comment (denyo) [reply packages/components/src/components/headline/headline.tsx]: Better would be to have a look at `link-social-styles.ts` and see how it it is solved there.
https://github.com/porsche-design-system/porsche-design-system/blob/issue/v3-alpha/packages/components/src/components/link-social/link-social-styles.ts

**Review by denyo — APPROVED** [review]

File: packages/components/src/components/heading/heading-styles.ts (hunk 7 above)
Comment (stephanschroeter) [reply packages/components/src/components/heading/heading-styles.ts]: we don't want to use variants at all for the new component.

File: packages/storefront/src/pages/utilities/scss/functions.md (hunk 21 above)
Comment (stephanschroeter) [reply packages/storefront/src/pages/utilities/scss/functions.md]: hehe, we'll have to update this section in general in the upcoming days since this package and docs are deprecated.
